### ISSUE TITLE
Server Error:  LiteException Cannot insert duplicate key in unique index

### ISSUE THREAD

**Issue by Elliander** [StrangeLoopGames/EcoIssues#8141]
While playing tonight, (EcoPC_v0.7.4.7-beta; Windows 10 64-bit) and running a server on the same machine, my computer froze up. As in, the note of a particular sound played forever and everything was frozen requiring a hard reboot. (My computer never freezes or crashes outside of using Eco and/or EcoServer, and I never seem to go above 25% CPU usage outside building a server and nothing overheats) When I pulled everything back up, the world loaded just fine. It was only around 5 minutes of lost work, so nothing serious there.

However, I noticed that with every block placed in the game client the Ecoserver would print the same line in red:

**Error writing to stat database: LiteException Cannot insert duplicate key in unique index '_id'. The duplicate value is '7873'.**

(I noticed because I keep the server text pulled up in one monitor, and the GUI on yet another monitor)

Interestingly, picking up a block didn't result in this error. I thought at first that maybe it was because I was dropping a block where I did before I crashed so I placed one elsewhere and that worked just fine. I logged out of the game client and logged back in, and everything was saved, but of course the server was still running so then I tried to save the world within the GUI and everything froze up. 

I closed it, and reopened, and then the server hung there using 0.2% of the CPU (which it randomly does for no reason, though more often when building a new server) so I closed and reopened the server and... everything looked OK. It was exactly how it was since the previous run of the server, and everything saved OK despite the error, but... the moment I placed a block down it said, once again:

 **Error writing to stat database: LiteException Cannot insert duplicate key in unique index '_id'. The duplicate value is '7873'.**

This could get very old very fast. I mean, everything appears to be working, but having the same error flood the logs repeatedly whenever someone drops a block isn't good. More importantly, it implies that something since the previous backup got corrupted, so I shut everything down to be safe. Oh, and this time it actually would save the world, so that is working now at least.

Is there any way to manually restore a particular backup? I mean, I figure that if I could load the backup before last I could see if that makes this error go away. 

Here's the full log from the first run after the crash:

> [04:32] Eco Server 0.7.4.7 beta
> [04:32] Server Initialization...
> [04:32] Starting LocalizationPlugin... Finished in 174.2ms
> [04:32] Starting ModKitPlugin... Finished in 143.9ms
> [04:33] Loading mods... Finished in 5.0ms
> [04:33] Starting NetworkManager... Finished in 34.5ms
> [04:33] Initializing items... Finished in 2,690.7ms
> [04:33] Starting WorldGeneratorPlugin... Finished in 62.0ms
> [04:33] Starting GameplayStatsPlugin... Finished in 0.1ms
> [04:33] Starting BackupPlugin... Finished in 9.1ms
> [04:33] Starting StorageManager... Finished in 2,605.7ms
> [04:35] Initializing skills... Finished in 64.3ms
> [04:35] Starting StatDatabase... Finished in 30.2ms
> [04:35] Starting WorldPlugin... Finished in 21,782.3ms
> [04:35] Loading World... Finished in 17,260.8ms
> [04:53] Initializing world... Finished in 4,475.2ms
> [04:57] Starting TimePlugin... Finished in 14.2ms
> [04:57] Loading Time... Finished in 0.6ms
> [04:57] Starting ReputationPlugin... Finished in 17.1ms
> [04:57] Loading Reputation... Finished in 5.5ms
> [04:57] Starting PlayerHousingManager... Finished in 0.7ms
> [04:57] Starting UserManager... Finished in 156.2ms
> [04:57] Loading Users... Finished in 108.2ms
> [04:57] Initializing inventories... Finished in 20.9ms
> [04:57] Starting EcoObjectManager... Finished in 39.0ms
> [04:57] Loading Objects... Finished in 27.4ms
> [04:57] Starting GarbagePlugin... Finished in 23.1ms
> [04:57] Starting ChatServer... Finished in 107.2ms
> [04:57] Starting DifficultySettings... Finished in 5.5ms
> [04:57] Starting ControllerManager... Finished in 1,545.8ms
> [04:57] Starting WirePlugin... Finished in 0.5ms
> [04:57] Starting EcoSim... Finished in 35,121.4ms
> [04:57] Starting PropertyManager... Finished in 79.1ms
> [04:57] Starting DisasterPlugin... Finished in 77.9ms
> [04:57] Loading Garbage... Finished in 1.1ms
> [04:57] Loading Property... Finished in 31.3ms
> [04:57] Loading Disasters... Finished in 0.1ms
> [04:57] Starting MinimapManager... Finished in 2.7ms
> [04:57] Loading Chat... Finished in 6.6ms
> [04:58] Loading Plants... Finished in 5,304.3ms
> [05:03] Initializing plants... Finished in 27,970.2ms
> [05:32] Starting RouteProbingPlugin... Finished in 6,814.0ms
> [05:32] Loading RouteProbing... Finished in 6,784.7ms
> [05:39] Starting WorldLayerManager... Finished in 16,468.7ms
> [05:39] Loading Layers... Finished in 1,785.5ms
> [05:41] Initializing world layers... Finished in 14,657.8ms
> [05:41] Loading ClimateSim... Finished in 0.4ms
> [05:56] Starting EconomyTracker... Finished in 1.4ms
> [05:56] Starting WorldObjectManager... Finished in 168.9ms
> [05:56] Loading WorldObjects... Finished in 91.4ms
> [05:56] Starting EconomyPlugin... Finished in 47.0ms
> [05:56] Loading Economy... Finished in 19.6ms
> [05:56] Starting RoomPlugin... Finished in 33.7ms
> [05:56] Loading Rooms... Finished in 5.1ms
> [05:56] Starting ObjectivePlugin... Finished in 27.7ms
> [05:56] Loading Objectives... Finished in 0.7ms
> [05:56] Starting Legislation... Finished in 284.1ms
> [05:56] Loading Laws... Finished in 6.6ms
> [05:56] Starting WebServerPlugin... Finished in 0.1ms
> [05:56] Starting HousingTierDefinition... Finished in 0.1ms
> [05:56] Initializing LocalizationPlugin... Finished in 65.4ms
> [05:56] Initializing WorldGeneratorPlugin... Finished in 0.3ms
> [05:56] Initializing UserManager... Finished in 18.0ms
> [05:56] Initializing MinimapManager... Finished in 4,410.6ms
> [06:01] Initializing WorldObjectManager... Finished in 97.8ms
> [06:01] Initializing RoomPlugin... Finished in 61.5ms
> [06:01] Initializing HousingTierDefinition... Finished in 0.3ms
> [06:01] Initializing WorldPlugin... Finished in 15,780.0ms
> [06:01] Ticking... Finished in 133.7ms
> [06:01] Compressing... Finished in 9,500.0ms
> [06:17] Initializing WorldLayerManager... Finished in 0.3ms
> [06:17] Initializing WebServerPlugin... Finished in 651.2ms
> [06:18] Server Initialization finished in 105,851.3ms
> [19:02] Error writing to stat database: LiteException Cannot insert duplicate key in unique index '_id'. The duplicate value is '7873'.
> [19:03] Error writing to stat database: LiteException Cannot insert duplicate key in unique index '_id'. The duplicate value is '7873'.
> [19:05] Error writing to stat database: LiteException Cannot insert duplicate key in unique index '_id'. The duplicate value is '7873'.
> [20:03] Error writing to stat database: LiteException Cannot insert duplicate key in unique index '_id'. The duplicate value is '7873'.
> [20:20] Error writing to stat database: LiteException Cannot insert duplicate key in unique index '_id'. The duplicate value is '7873'.
> [20:55] Error writing to stat database: LiteException Cannot insert duplicate key in unique index '_id'. The duplicate value is '7873'.
> [20:56] Error writing to stat database: LiteException Cannot insert duplicate key in unique index '_id'. The duplicate value is '7873'.
> [20:57] Error writing to stat database: LiteException Cannot insert duplicate key in unique index '_id'. The duplicate value is '7873'.
> [20:59] Error writing to stat database: LiteException Cannot insert duplicate key in unique index '_id'. The duplicate value is '7873'.
> [21:04] Error writing to stat database: LiteException Cannot insert duplicate key in unique index '_id'. The duplicate value is '7873'.
> [21:05] Error writing to stat database: LiteException Cannot insert duplicate key in unique index '_id'. The duplicate value is '7873'.
> [21:07] Error writing to stat database: LiteException Cannot insert duplicate key in unique index '_id'. The duplicate value is '7873'.
> [21:08] Error writing to stat database: LiteException Cannot insert duplicate key in unique index '_id'. The duplicate value is '7873'.
> [21:12] Error writing to stat database: LiteException Cannot insert duplicate key in unique index '_id'. The duplicate value is '7873'.
> [21:21] Error writing to stat database: LiteException Cannot insert duplicate key in unique index '_id'. The duplicate value is '7873'.
> [23:04] Error writing to stat database: LiteException Cannot insert duplicate key in unique index '_id'. The duplicate value is '7873'.
> [23:08] Error writing to stat database: LiteException Cannot insert duplicate key in unique index '_id'. The duplicate value is '7873'.

This is from the last run:

> 
> [33:47] Eco Server 0.7.4.7 beta
> [33:47] Server Initialization...
> [33:47] Starting LocalizationPlugin... Finished in 199.0ms
> [33:47] Starting ModKitPlugin... Finished in 109.7ms
> [33:47] Loading mods... Finished in 4.1ms
> [33:47] Starting NetworkManager... Finished in 28.4ms
> [33:47] Initializing items... Finished in 2,288.6ms
> [33:47] Starting WorldGeneratorPlugin... Finished in 47.3ms
> [33:47] Starting GameplayStatsPlugin... Finished in 0.8ms
> [33:47] Starting BackupPlugin... Finished in 2.5ms
> [33:47] Starting StorageManager... Finished in 1,776.1ms
> [33:49] Initializing skills... Finished in 67.0ms
> [33:49] Starting StatDatabase... Finished in 23.5ms
> [33:49] Starting WorldPlugin... Finished in 22,440.6ms
> [33:50] Loading World... Finished in 15,183.3ms
> [34:05] Initializing world... Finished in 7,212.8ms
> [34:12] Starting TimePlugin... Finished in 14.6ms
> [34:12] Loading Time... Finished in 0.7ms
> [34:12] Starting ReputationPlugin... Finished in 17.2ms
> [34:12] Loading Reputation... Finished in 5.4ms
> [34:12] Starting PlayerHousingManager... Finished in 0.7ms
> [34:12] Starting UserManager... Finished in 137.9ms
> [34:12] Loading Users... Finished in 90.7ms
> [34:12] Initializing inventories... Finished in 20.4ms
> [34:12] Starting EcoObjectManager... Finished in 70.7ms
> [34:12] Loading Objects... Finished in 58.9ms
> [34:12] Starting ControllerManager... Finished in 1,379.7ms
> [34:12] Starting GarbagePlugin... Finished in 35.9ms
> [34:12] Starting DifficultySettings... Finished in 7.4ms
> [34:12] Starting WirePlugin... Finished in 1.3ms
> [34:12] Starting PropertyManager... Finished in 89.3ms
> [34:12] Starting EcoSim... Finished in 34,304.7ms
> [34:12] Starting ChatServer... Finished in 113.7ms
> [34:12] Starting DisasterPlugin... Finished in 87.5ms
> [34:12] Loading Garbage... Finished in 1.7ms
> [34:12] Loading Property... Finished in 33.8ms
> [34:12] Loading Disasters... Finished in 0.1ms
> [34:12] Starting MinimapManager... Finished in 3.5ms
> [34:12] Loading Chat... Finished in 6.4ms
> [34:12] Loading Plants... Finished in 4,423.6ms
> [34:17] Initializing plants... Finished in 27,263.8ms
> [34:46] Starting RouteProbingPlugin... Finished in 8,203.6ms
> [34:46] Loading RouteProbing... Finished in 8,179.6ms
> [34:55] Starting WorldLayerManager... Finished in 20,270.8ms
> [34:55] Loading Layers... Finished in 2,057.2ms
> [34:57] Initializing world layers... Finished in 18,188.8ms
> [34:57] Loading ClimateSim... Finished in 0.4ms
> [35:15] Starting EconomyTracker... Finished in 4.0ms
> [35:15] Starting WorldObjectManager... Finished in 139.5ms
> [35:15] Loading WorldObjects... Finished in 93.1ms
> [35:15] Starting EconomyPlugin... Finished in 33.7ms
> [35:15] Loading Economy... Finished in 18.9ms
> [35:15] Starting RoomPlugin... Finished in 24.5ms
> [35:15] Loading Rooms... Finished in 6.0ms
> [35:15] Starting ObjectivePlugin... Finished in 21.5ms
> [35:15] Loading Objectives... Finished in 0.6ms
> [35:15] Starting Legislation... Finished in 276.5ms
> [35:15] Loading Laws... Finished in 7.5ms
> [35:15] Starting WebServerPlugin... Finished in 0.4ms
> [35:15] Starting HousingTierDefinition... Finished in 0.1ms
> [35:15] Initializing LocalizationPlugin... Finished in 53.6ms
> [35:16] Initializing WorldGeneratorPlugin... Finished in 0.6ms
> [35:16] Initializing UserManager... Finished in 44.4ms
> [35:16] Initializing MinimapManager... Finished in 2,254.0ms
> [35:18] Initializing WorldObjectManager... Finished in 110.7ms
> [35:18] Initializing RoomPlugin... Finished in 60.8ms
> [35:18] Initializing HousingTierDefinition... Finished in 0.3ms
> [35:18] Initializing WorldPlugin... Finished in 15,636.0ms
> [35:18] Ticking... Finished in 138.2ms
> [35:18] Compressing... Finished in 8,749.4ms
> [35:34] Initializing WorldLayerManager... Finished in 0.3ms
> [35:34] Initializing WebServerPlugin... Finished in 533.2ms
> [35:35] Server Initialization finished in 107,981.5ms
> [37:31] Error writing to stat database: LiteException Cannot insert duplicate key in unique index '_id'. The duplicate value is '7873'.
> [40:11] Saving world... Finished in 2,824.6ms

Are there spoiler tags? This seems a bit long.

**Comment by Elliander** [comment]
I was able to solve the issue: First I backed up the entire directory structure for the server manually (just in case) and then I deleted 

Storage/Game.db

replacing it with the most recent backup:

Storage/Game-2018-06-24-21.15.26/Game.db

But I kept the Game.eco file as it was, since that backup was a few hours ago and didn't want to lose all of my work.

After I loaded the server everything was where it should be, and I double checked stockpiles and items are all accounted for there as well, and there are no errors when dropping items. I can share any of the files from that original though if needed to track the problem down.
